### DISCUSSION TITLE
The specified ibrav 14 is not supported

### DISCUSSION THREAD

**Discussion by songxq1994** [SSCHAcode/python-sscha#102] [Q&A] [answered]
Dear SSCHA developers, 

I am very interested in SSCHA software. Recently, I have met such a problem about symmetry, as below:
-----------------------------------------------------------------------------  
raise ValueError("Error, the specified ibrav %d is not supported." % ibrav)  
ValueError: Error, the specified ibrav 14 is not supported.
-----------------------------------------------------------------------------   
I am researching some low-symmetry structures, but the calculations of _Triclinic_ crystal are likely not supported in the SSCHA code.    
Are the restrictions originated from Quantum Espresso or SSCHA?  
Is this based on physical or procedural considerations?
What can I do to resolve this issue?   

Best regards,   
X. Song.

**Comment by mesonepigreco** [comment, marked as answer]
Hi, I tried to implement ibrav 14 right now, can you update cellconstructor to the last version in the master and try again? It would be good if you can check if it is correct by visualizing the structure by running the following python code (replace dynmat with the path to your espresso dynamical matrix (excluding the final number):
```
import ase, ase.visualize, cellconstructor as CC, cellconstructor.Phonons
ase.visualize.view(CC.Phonons.Phonons("dynmat").structure.get_ase_atoms())
```

Otherwise, you could use ibrav 0 which allow you to specify all the atomic position and lattice version directly, which is the safest option for the SSCHA code.

Let us know if the implementation works,
bests,
Lorenzo

**Reply by songxq1994** [reply]
Hello, Lorenzo,

Thanks for your quick response. I have solved this problem now based on your effective suggestions using ibrav 0 in dynmat file.

**Please allow me to provide more details about this solution and help other users to deal with this problem.**
When we use ibrav=14, the format of dynmat files are 

```
Dynamical matrix file 
Electron-phonon coefficients for H2
  1   24 14  9.5112184  0.5716608  0.8448067  0.0779002 -0.7848733  0.0017269
           1  'H   '    918.735796069331
Atomic positions
...
Dynamical Matrix
...
```

When we use ibrav=0，the format of dynmat files can be adjust to


```
Dynamical matrix file
Electron-phonon coefficients for H2
  1   24 0  9.5512184  0.0000000  0.0000000  0.0000000  0.0000000   0.0000000
Basic vectors
      1.000000000    0.000000000    0.000000000
      0.000987201    0.571659998    0.000000000
     -0.663066262    0.066955746    0.519190223
           1  'H   '    918.735796069331
Atomic positions
...
Dynamical Matrix
...
```

Therefore, only the first few lines need be modified in dynmat files so that we don't need to recalculate dynmat file.

### The next problem is about the input of number of dynamical matrix.

```
#Let us load the starting dynamical matrix
dyn = CC.Phonons.Phonons("matdyn", nqirr = 36)
```

When I use the 4 $\times$ 4 $\times$ 4 $q$-mesh (36 irreducible $q$-point), the error is "double free or corruption (out)". Then I change the nqirr to a small number, e.g. 1 or 3, the code works well. 

My question is if I use a small $q$-mesh (e. g. 2 $\times$ 2 $\times$ 2) to obtain initial dynamical matrix in QE and carry out SSCHA code to obtain the new dynamical matrix considering anharmonic and nuclear quantum effect, whether the $q$-mesh is sufficient for the subsequent following electron-phonon coupling calculation?

Bests,

Xianqi.

**Reply by mesonepigreco** [reply]
Hi, thanks for sharing your solution, it will be useful also to other users.

Regarding the second point, it is always better to start the SSCHA calculation on a lower q-mesh (if the anharmonicity is converged in that cell), then, if you need dynamical matrices on a denser grid, you can interpolate. 

This error that you are showing seems an out-of-memory error (the code is trying to allocate too much memory to store all the dynamical matrices together. How big (number of atoms) is your system?

**Reply by songxq1994** [reply]
>  if the anharmonicity is converged in that cell, then, if you need dynamical matrices on a denser grid, you can interpolate.

Could you give me more detailed advice about how to judge the convergence of anharmonicity and how to interpolate for obtaining dynamical matrices on a denser grid based on the SSCHA results?

> How big (number of atoms) is your system?

My simulated structure contains 24 atoms with space group _P_-1 and I allocated 4GB RAM to the ubuntu virtual machine on my PC.

**Reply by mesonepigreco** [reply]
Hi, it is very difficult to judge if the anharmonicity is well converged. In principle, one should study how the observable that you want to study changes as a function of the mesh grid but, in practice, simulating too large cells could be really challenging. In general, supercells containing about 100 atoms are good to converge, so your 2x2x2 mesh with 24 atoms contains already 192 atoms which should be enough. 

To interpolate, you can load the dynamical matrix:
```
import cellconstructor as CC, cellconstructor.Phonons
dyn_222 = CC.Phonons.Phonons("my_sscha_dyn", nqirr = X) # X is the nqirr of the 2x2x2 mesh
new_dyn_444 = dyn.Interpolate((2,2,2), (4,4,4))  # Interpolate the dynamical matrix on a 4x4x4 mesh
new_dyn_444.save_qe("my_new_sscha_dyn") # The final SSCHA dynamical matrix on a 4x4x4
```

I do not understand why you are running out of memory. 4 Gb of RAM should be enough even to store your dynamical matrix in real space. Does the code crashes already when loading the dynamical matrix?

**Reply by songxq1994** [reply]
Hi, Lorenzo,
Sorry for my late response. Thanks for your detailed reply. I have not met similar problems again about running out of memory. 
> Does the code crashes already when loading the dynamical matrix?

The code is terminated immediately without any data interchange between PC and clusters.
#-----
I tried my best to perform the calculations of Raman, IR, and Superconductively (SC), which have related descriptions in the discussions "Infrared Response from SSCHA? #46 and Superconductivity calculations with phonons obtained from SSCHA code #103" but I still met some questions as below:

**For SC calculations:** 
(1) I carried out SSCHA to get the new structure and anharmonic matdyn* files;  
(2) I performed electron and phonon calculations using the new structures and retained the files " ./_ph0/prefix.dv, ./_ph0/prefix.q_\*, and ./_ph0/prefix.phsave/patterns\* "; 
(3) I replaced the harmonic matdyn* files with anharmonic matdyn* files from SSCHA and calculated the e-p coupling and Tc in QE code. 
The error in ph.out is 
```
 Reading dVscf from file dv   
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%     
 Error in routine readmat (2):
     inconsistent data 
```
**For the calculations of IR and Raman:**
>In #46, the easiest way is to perform a computation of the dielectric tensor and effective charges within a DFT code and copy them inside the first dynamical matrix file (as done in the usual output of quantum espresso). Then you can use the quantum espresso dynmat.x code on the converged SSCHA dynamical matrix to get IR activity of the modes.

My questions are: 
(1) Only NC and LDA pseudopotential are likely supported for the calculations of IR and Raman in QE code. How should I handle the results based on the USPP and GGA pseudopotential; 
(2) Does it mean that I perform a computation of the dielectric tensor and effective charges **using the new structures from SSCHA** and copy them inside the first **anharmonic SSCHA** dynamical matrix file?


Best Wishes,

Xianqi.

**Reply by mesonepigreco** [reply]
Sorry for my late reply, since there was a marked answer I did not get the notification. 
Anyway, there is a new discussion for how to compute superconductivity at  #103 , I suggest you can look at that.

Regarding the IR and Raman, yes, you should use NC pseudos with LDA to get the Raman tensor. However, usually, this is a very good approximation for the Raman intensities. For IR, effective charges and the dielectric tensor should also be implemented for USPP and GGA. 
You can use the structure from the SSCHA and copy effective charges, dielectric tensor, and Raman tensor inside the last anharmonic SSCHA dynamical matrix file.

**Reply by songxq1994** [reply]
Hi, 
Thanks for your kind reply and helpful hints.  I will continue these tests and provide further feedback.

Bests,
Xianqi.
